src/Lib.P2P/SecureCommunication/Secio1.cs

```diff
@@ -0,0 +1,246 @@
+            // step 1.1 Identify -- get identity from their key
+            var remoteProposal =
+                Serializer.DeserializeWithLengthPrefix<Secio1Propose>(stream, PrefixStyle.Fixed32BigEndian);
+            var ridAlg = remoteProposal.PublicKey.Length <= 48 ? "identity" : "sha2-256";
+            var remoteId = MultiHash.ComputeHash(remoteProposal.PublicKey, ridAlg);
+            if (remotePeer.Id == null)
+            {
+                remotePeer.Id = remoteId;
+            }
+            
+            else if (remoteId != remotePeer.Id)
+            {
+                throw new Exception($"Expected peer '{remotePeer.Id}', got '{remoteId}'");
+            }
+
+            // =============================================================================
+            // step 1.2 Selection -- select/agree on best encryption parameters
+            // to determine order, use cmp(H(remote_pubkey||local_rand), H(local_pubkey||remote_rand)).
+            //   oh1 := hashSha256(append(proposeIn.GetPubkey(), nonceOut...))
+            //   oh2 := hashSha256(append(myPubKeyBytes, proposeIn.GetRand()...))
+            //   order := bytes.Compare(oh1, oh2)
+            byte[] oh1;
+            byte[] oh2;
+            using (var hasher = MultiHash.GetHashAlgorithm())
+            await using (var ms = new MemoryStream())
+            {
+                ms.Write(remoteProposal.PublicKey, 0, remoteProposal.PublicKey.Length);
+                ms.Write(localProposal.Nonce, 0, localProposal.Nonce.Length);
+                ms.Position = 0;
+                oh1 = hasher.ComputeHash(ms);
+            }
+
+            using (var hasher = MultiHash.GetHashAlgorithm())
+            await using (var ms = new MemoryStream())
+            {
+                ms.Write(localProposal.PublicKey, 0, localProposal.PublicKey.Length);
+                ms.Write(remoteProposal.Nonce, 0, remoteProposal.Nonce.Length);
+                ms.Position = 0;
+                oh2 = hasher.ComputeHash(ms);
+            }
+
+            var order = 0;
+            for (var i = 0; order == 0 && i < oh1.Length; ++i) order = oh1[i].CompareTo(oh2[i]);
+            if (order == 0)
+            {
+                throw new Exception("Same keys and nonces; talking to self");
+            }
+            
+            var curveName = SelectBest(order, localProposal.Exchanges, remoteProposal.Exchanges);
+            if (curveName == null)
+            {
+                throw new Exception("Cannot agree on a key exchange.");
+            }
+
+            var cipherName = SelectBest(order, localProposal.Ciphers, remoteProposal.Ciphers);
+            if (cipherName == null)
+            {
+                throw new Exception("Cannot agree on a chipher.");
+            }
+
+            var hashName = SelectBest(order, localProposal.Hashes, remoteProposal.Hashes);
+            if (hashName == null)
+            {
+                throw new Exception("Cannot agree on a hash.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)